thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java

```diff
@@ -0,0 +1,320 @@
+        // Double check with data source as 2 things are possible.
+        // 1. This dataset/source may not support availability events
+        // 2. The data availability event pipeline has some issue.
+        // We want to measure the overall dataset availability (filters are not taken into account)
+        MetricSlice metricSlice = MetricSlice.from(me.getId(), info.getStart(), info.getEnd(), ArrayListMultimap.<String, String>create());
+        DataFrame dataFrame = this.provider.fetchTimeseries(Collections.singleton(metricSlice)).get(metricSlice);
+        if (dataFrame == null || dataFrame.isEmpty()) {
+          // no data
+          if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+            createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+          }
+        } else {
+          datasetLastRefreshTime = dataFrame.getDoubles("timestamp").max().longValue();
+          if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+            if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+              createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+            }
+          }
+        }
+      } else if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+        // Optimize for the common case - the common case is that the data availability events are arriving
+        // correctly and we need not re-fetch the data to double check.
+        if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+          createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error(String.format("Failed to run sla check on metric URN %s", me.getUrn()), e);
+    }
+  }
+
+  /**
+   * We say the data is missing we do not have data in the sla detection window.
+   * Or more specifically if the dataset watermark is below the sla detection window.
+   */
+  private boolean isMissingData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info) {
+    return datasetLastRefreshTime < info.getStart();
+  }
+
+  /**
+   * We say the data is partial if we do not have all the data points in the sla detection window.
+   * Or more specifically if we have at least 1 data-point missing in the sla detection window.
+   *
+   * For example:
+   * Assume that the data is delayed and our current sla detection window is [1st Feb to 3rd Feb). During this scan,
+   * let's say data for 1st Feb arrives. Now, we have a situation where partial data is present. In other words, in the
+   * current sla detection window [1st to 3rd Feb) we have data for 1st Feb but data for 2nd Feb is missing. This is the
+   * partial data scenario.
+   */
+  private boolean isPartialData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info, DatasetConfigDTO datasetConfig) {
```

Comment:
Why we care about partial data? In our current setting the endTime of detection window is current time, and it is very common that we have partial data.

Reply:
discussed offline! others - pls refer to the comment in the code.